`ros2 topic echo` deserialization failed using connext

## Bug report

**Required Info:**

- Operating System:
  - Windows 10
- Installation type:
  - binaries
- Version or commit hash:
  - https://ci.ros2.org/view/packaging/job/packaging_windows/1046/
- DDS implementation:
  - RTI Connext
- Client library (if applicable):
  - N/A

#### Steps to reproduce issue

Start a talker executable in one terminal
```bat
set RMW_IMPLEMENTATION=rmw_connext_cpp
ros2 run demo_nodes_cpp talker
```

Echo the topic in another terminal
```bat
set RMW_IMPLEMENTATION=rmw_connext_cpp
REM make sure the daemon uses connext
ros2 daemon stop
ros2 daemon start

ros2 topic echo /chatter
```

#### Expected behavior

Every message published would be echoed on  the console

#### Actual behavior

The first message is printed, but then an error message `deserialize from cdr buffer failed` appears. The command never prints another message despite the talker continuing to publish.

```bat
C:\dev\ros2-windows>ros2 topic echo /chatter
RTI Data Distribution Service Evaluation License issued to OSRF (OSRF01) dthomas@osrfoundation.org For non-production use only.
Expires on 5-Nov-2018 See www.rti.com for more information.
data: 'Hello World: 50'

deserialize from cdr buffer failed
```

#### Additional information
Additionally, the command `ros2 topic echo /chatter` cannot be interrupted using `CTRL` + `c`. I had to kill the process in the task manager.

I've seen this error also get printed in our nightlies e.g.
```
2:  - /home/rosbuild/ci_scripts/ws/build/test_communication/test_publisher_subscriber_cpp BoundedArrayNested
2: cdr stream doesn't contain data
2: deserialize from cdr buffer failed
2: [ERROR] [rclcpp]: could not deserialize serialized message on topic '/test_message_BoundedArrayNested': can't convert cdr stream to ros message, at /home/rosbuild/ci_scripts/ws/src/ros2/rmw_connext/rmw_connext_cpp/src/rmw_take.cpp:185, at /home/rosbuild/ci_scripts/ws/src/ros2/rcl/rcl/src/rcl/subscription.c:258
```
(appears 17 times in various tests in https://ci.ros2.org/view/nightly/job/nightly_linux_release/854/consoleFull, for example)

--- a/ros2topic/ros2topic/verb/echo.py
+++ b/ros2topic/ros2topic/verb/echo.py
@@ -15,14 +15,13 @@
 from argparse import ArgumentTypeError
 from collections import OrderedDict
 import importlib
-import os
 import sys
 
 import rclpy
 from rclpy.expand_topic_name import expand_topic_name
 from rclpy.qos import qos_profile_sensor_data
 from rclpy.validate_full_topic_name import validate_full_topic_name
-from ros2cli.node.direct import DirectNode, HIDDEN_NODE_PREFIX
+from ros2cli.node.direct import DirectNode
 from ros2topic.api import get_topic_names_and_types
 from ros2topic.api import TopicNameCompleter
 from ros2topic.verb import VerbExtension
@@ -77,14 +76,8 @@ def main(args):
         callback = subscriber_cb(args)
     else:
         callback = subscriber_cb_csv(args)
-
-    rclpy.init()
-
-    node_name_suffix = '_%d' % os.getpid()
-    node = rclpy.create_node(HIDDEN_NODE_PREFIX + 'ros2topic_echo_node' + node_name_suffix)
-    subscriber(node, args.topic_name, args.message_type, callback)
-    node.destroy_node()
-    rclpy.shutdown()
+    with DirectNode(args) as node:
+        subscriber(node, args.topic_name, args.message_type, callback)
 
 
 def register_yaml_representer():

--- a/ros2topic/ros2topic/verb/echo.py
+++ b/ros2topic/ros2topic/verb/echo.py
@@ -77,7 +77,7 @@ def main(args):
     else:
         callback = subscriber_cb_csv(args)
     with DirectNode(args) as node:
-        subscriber(node, args.topic_name, args.message_type, callback)
+        subscriber(node.node, args.topic_name, args.message_type, callback)
 
 
 def register_yaml_representer():

--- a/ros2topic/ros2topic/verb/echo.py
+++ b/ros2topic/ros2topic/verb/echo.py
@@ -15,13 +15,14 @@
 from argparse import ArgumentTypeError
 from collections import OrderedDict
 import importlib
+import os
 import sys
 
 import rclpy
 from rclpy.expand_topic_name import expand_topic_name
 from rclpy.qos import qos_profile_sensor_data
 from rclpy.validate_full_topic_name import validate_full_topic_name
-from ros2cli.node.direct import DirectNode
+from ros2cli.node.direct import DirectNode, HIDDEN_NODE_PREFIX
 from ros2topic.api import get_topic_names_and_types
 from ros2topic.api import TopicNameCompleter
 from ros2topic.verb import VerbExtension
@@ -76,8 +77,14 @@ def main(args):
         callback = subscriber_cb(args)
     else:
         callback = subscriber_cb_csv(args)
-    with DirectNode(args) as node:
-        subscriber(node, args.topic_name, args.message_type, callback)
+
+    rclpy.init()
+
+    node_name_suffix = '_%d' % os.getpid()
+    node = rclpy.create_node(HIDDEN_NODE_PREFIX + 'ros2topic_echo_node' + node_name_suffix)
+    subscriber(node, args.topic_name, args.message_type, callback)
+    node.destroy_node()
+    rclpy.shutdown()
 
 
 def register_yaml_representer():

